File: tmp/sccs2rcs.py
#!/usr/bin/env python
#
# sccs2rcs is a script to convert an existing SCCS history into an RCS
# history without losing any of the information contained therein.
#
# This tool has a complicated history for its size.  It began life as a 
# script in C-shell written by a hacker named Ken Cox in 1991.
#
# In 1993 I applied some minor fixes for newer RCS versions and it 
# shipped as part of one of Rich Morin's Prime Time Freeware releases.
#
# In 1994 Brian Berliner hacked Ken Cox's original in a slightly
# different direction than mine, adding more SCCS keyword translations
# and an improvement in dealing with branches.  At least, Brian *thinks*
# it was 1994...
#
# In 2002 I fixed it not to require the Berkeley sccs(1) front end.
# In 2003 I folded in a minor bug fix from Roger Pilkey.
# In 2007 I merged in the missing features from Brian Berliner's version
# and sent it to the GNU RCS project hoping they'd take it off my hands.
# 
# Then it started to bother me that I was casting the thing loose in a
# dying language, so I decided to port it to Python and clean it up.
# I added error checking, added the comments-file processing, added
# auto-checkout of RCS workfiles replacing SCCS ones, and removed
# no-longer-needed options. Finally I made it able to parse tag data from
# snapshots made by Emacs VC mode.
#
# Ken said:
# "This file may be copied, processed, hacked, mutilated, and
# even destroyed as long as you don't tell anyone you wrote it."
# Those were more innocent days.  This version is released under MIT terms. 
#
# Ken's things to note:
#   + It will NOT delete or alter your ./SCCS history under any circumstances.
#
#   + Run in a directory where ./SCCS exists and where you can create ./RCS
#
#   + Date, time, author, comments, branches, are all preserved.
#
#   + If a command fails somewhere in the middle, it bombs with
#     a message -- remove what it's done so far and try again.
#     (`rm -rf RCS; unget SCCS/s.*' should do the job.)
#     There is no recovery and exit is far from graceful.
#     If a particular module is hanging you up, consider
#     doing it separately; move it from the current area so that
#     the next run will have a better chance or working.
#     Also (for the brave only) you might consider hacking
#     the s-file for simpler problems:  I've successfully changed
#     the date of a delta to be in sync, then run "sccs admin -z"
#     on the thing [to regenerate its checksum].
#
# Eric's history:
#
# 1.1: I added some portability fixes and the -u option.  More importantly,
# I removed the dependencies on the BSD/SV `sccs' utility; this script
# will now work on any system that has the base SCCS/RCS tools available.
#
# 1.2: Merges in Roger Pilkey's fix for the case in which date spans 
# cross the millenium boundary.  Use GNU tac instead of tail, which can't 
# reverse large files properly.
#
# 1.3: Merges in Brian Berliner's branch hack from the other fork
# of this script, the one distributed with CVS.  That and four SCCS 
# keyword substitutions were we only thing it did that this didn't.
#
# 1.4: Python rewrite describe aboved.  Interface cleanup.  Auto-checkout
# of RCS workfiles corresponding to SCCS workfiles present when started.
# Lifting of VC mode snapshot info.  Initial-comments file.
#
# 1.5: Bug fix in comment extraction from Peter Grandcourt and Frederic
# G. Marand.
#
# Eric S. Raymond <esr@thyrsus.com>
#
# $Id: sccs2rcs 36 2007-09-14 20:04:00Z esr $

import os, sys, getopt, re, commands

# Here are the translations for SCCS % keywords to RCS-style $ keywords.
# The quotes surround the dollar signs are to fool RCS and later version-
# control systems when the script is checked in.  In the unlikely event
# that you need more keyword translations, add them to the end of this list.
# The ones this misses are %[ABCDFHLMPQRSTYZ]%; the script will issue
# warnings if it sees them after translation.
conversions = [
    ('%W%[ \t]*%G%',			"$""Id""$"),
    ('%W%[ \t]*%E%',			"$""Id""$"),
    ('%W%',				"$""Id""$"),
    ('%Z%%M%[ \t]*%I%[ \t]*%G%',	"$""SunId""$"),
    ('%Z%%M%[ \t]*%I%[ \t]*%E%',	"$""SunId""$"),
    ('%M%[ \t]*%I%[ \t]*%G%',		"$""Id""$"),
    ('%M%[ \t]*%I%[ \t]*%E%',		"$""Id""$"),
    ('%M%',				"$""RCSfile""$"),
    ('%I%',				"$""Revision""$"),
    ('%G%',				"$""Date""$"),
    ('%E%',				"$""Date""$"),
    ('%U%',				""),
    ]
conversions = map(lambda (s, t): (re.compile(s), t), conversions)

months = ("Jan","Feb","Mar","Apr","May","Jun",
          "Jul","Aug","Sep","Oct","Nov","Dec")

def writeable(f):
    return os.access(f, os.W_OK)

def sccsmaster(f):
    return os.path.join("SCCS", "s." + f)

def is_sccsmaster(f):
    return os.path.basename(f).startswith("s.")
    
def workfile(f):
    assert os.path.basename(f).startswith("s.")
    return os.path.basename(f)[2:]

def rcsfile(f):
    return os.path.join("RCS", f + ",v")

def record(msg):
    global logtext
    msg = "sccs2rcs: " + msg
    logtext += msg + "\n"

def progress(msg):
    global logtext
    msg = "sccs2rcs: " + msg
    logtext += msg + "\n"
    print msg

def do_it(command):
    global logtext
    logtext += "$ " + command + "\n"
    (st, out) = commands.getstatusoutput(command)
    logtext += out + "\n"
    return out + "\n"

def die(msg):
    print >>sys.stderr, "sccs2rcs: fatal error, ", msg
    record("error recovery begins")
    for master in os.listdir("SCCS"):
        if is_sccsmaster(master):
            do_it("unget " + master)
    print >>sys.stderr, """\
Danger!  Danger!
Incomplete history in ./RCS -- remove it.
Original unchanged history in ./SCCS.
"""
    if logtext:
        print "Log follows:"
        print logtext
        print "Log ends here."
    sys.exit(1)

def warn(msg):
    global logtext
    msg = "sccs2rcs: warning, " + msg
    logtext += msg + "\n" 
    print >>sys.stderr, msg

def do_it_or_die(command):
    global logtext
    logtext += "$ " + command + "\n"
    (st, out) = commands.getstatusoutput(command)
    logtext += out + "\n"
    if st != 0:
        die("'%s' failed" % command)
    return out + "\n"

if __name__ == '__main__':
    comments = None
    (options, arguments) = getopt.getopt(sys.argv[1:], "c:")
    for (switch, val) in options:
        if switch == '-c':
            comments = val

    # User may have supplied a comments file
    initstrings = {} 
    if comments:
        for line in open(comments):
            try:
                line = line.strip()
                if not line:	# ignore blank lines
                    continue
                (filename, initstring) = line.split(":")
            except ValueError:
                die("ill-formed line '%s' in comments file" % line)
            if filename in initstrings:
                die("duplicate filename in initstrings")
            initstrings[filename] = initstring.strip()

    # If the CSSC package is installed, the SCCS commands might be here.
    # This assumption works under Debian/Ubuntu and will be harmless elsewhere.
    os.environ["PATH"] += ":/usr/lib/cssc"

    logtext = "" 

    # Sanity checks on the directory
    if not writeable("."):
        die("./ is not writeable by you.")
    elif not os.path.isdir("SCCS"):
        die("no SCCS directory to convert.")
    for filename in os.listdir("SCCS"):
        if filename.startswith("s.") and writeable(filename):
            die(filename + " is locked for edit...unlock before converting.")
    if os.path.isdir("RCS"):
        warn("RCS directory already exists.")
        if len(os.listdir("RCS")) > 2:	# Allow . and ..
               die("RCS directory is nonempty.")
    else:
        try:
            os.mkdir("RCS")
        except OSError:
            die("failed while attempting to create RCS directory.")

    # Loop over every s-file in SCCS dir
    workfiles = []
    for master in filter(lambda f: f.startswith("s."), os.listdir("SCCS")):
        stem = master[2:]
        firsttime = True

        # These are the files we'll need to restore later
        if os.path.exists(stem):
            workfiles.append(stem)

        # Extract a list of revisions from the master 
        revisions = []
        try:
            for line in os.popen("prs " + sccsmaster(stem), "r"):
                if line.startswith("D "):
                    revisions.append(line.split()[1])
        except OSError:
            die("open of master file %s failed" % master)
        except IndexError:
            die("garbled revision line in master file " + master)
        revisions.reverse()

        # Now use prs to run through the revision list extracting data
        record("%s revisions: %s" % (stem, ", ".join(revisions)))
        for rev in revisions:
            cmd = 'prs -d":Dd: :Dm: :Dy: :T: :P:" -r%s SCCS/s.%s' % (rev, stem)
            date = do_it_or_die(cmd).strip()
            try:
                (day, month, year, tod, author) = date.split()
            except ValueError:
                die("prs output %s from '%s'was ill-formed" % (`date`, cmd))

            # We assume here that no SCCS archives were made before 1970.
            # If you're converting an SCCS archive made after 2070, you lose.
            if int(year) < 70:
                year = "20" + year
            else:
                year = "19" + year

            rcsdate = day + " " + months[int(month)-1] + " " + year + " " + tod
            progress("==> file %s, rev=%s, date=%s, author=%s" \
                  % (stem, rev, rcsdate, author)) 
            do_it_or_die("get -e -r%s %s" % (rev, sccsmaster(stem)))
            record("checked out of SCCS")

            # replace SCCS keywords with RCS keywords
            try:
                rfp = open(stem, "r")
                stemtext = rfp.read()
                rfp.close()
                for (regexp, rcskey) in conversions:
                    stemtext = regexp.sub(rcskey, stemtext)
                wfp = open(stem, "w")
                wfp.write(stemtext)
                wfp.close()
            except OSError:
                    die("keyword substitution failed")
            record("performed keyword substitutions")

            # check file into RCS
            if firsttime:
                firsttime = False
                if stem in initstrings:
                    cmd = 'ci -f -r%s -d"%s" -w%s -t-"%s" %s' \
                          % (rev, rcsdate, author, initstrings[stem], stem)
                else:
                    if initstrings:
                        warn("no description for %s in comments file" % stem)
                    cmd = 'ci -f -r%s -d"%s" -w%s -t/dev/null %s' \
                          % (rev, rcsdate, author, stem)
                cmd = "echo 'Initial revision.' | " + cmd
                do_it_or_die(cmd)
                record("initial checkin successful")
            else:
                # get RCS lock    
                revparts = rev.split(".")
                lckrev = ".".join(revparts[:-1])
                if rev.count(".") > 1:
                    # need to lock the branch -- it is OK if the lock fails
                    #
                    # I asked Brian and he explained this as follows:
                    # "As it turns out, with RCS, if you are adding the
                    # first revision of a new branch, you don't need
                    # to lock the trunk. In fact, you don't want to
                    # lock the trunk because checking in the revision
                    # on the branch will not unlock the trunk. So, the
                    # hack I applied was to get it to add branches
                    # correctly."
                    do_it("rcs -l%s %s" % (lckrev, stem))
                    record("got branch lock")
                else:
                    # need to lock the trunk -- must succeed
                    do_it_or_die("rcs -l %s" % (stem,))
                    record("got trunk lock")
                # Extract the delta message for this revision
                deltamsg = do_it_or_die("prs -r%s SCCS/s.%s" % (rev, stem))
                deltalines = deltamsg.split("\n")
                while deltalines:
                    line = deltalines[0]
                    deltalines.pop(0)
                    if line.startswith("COMMENTS"):
                        break
                # Strip ;eading blank lines if present.
                if deltalines != ['', '']:
                    while not deltalines[0].strip():
                        deltalines.pop(0)
                deltamsg = "\n".join(deltalines)
                cmd = 'ci -f -r%s -d"%s" -w%s %s' % (rev, rcsdate, author, stem)
                # FIXME: output of this command isn't logged
                record("$ " + cmd)
                try:
                    cfp = os.popen(cmd, "w")
                    cfp.write(deltamsg)
                    cfp.close()
                except OSError:
                    die("RCS checkin of delta failed")
            # We're done with this master, unlock it.
	    do_it_or_die("unget -r%s %s" % (rev, sccsmaster(stem)))

    # Generate tags from snapshots created by Emacs VC mode
    tagnames = os.path.join("SCCS", "VC-names")
    if os.path.exists(tagnames):
        for line in open(tagnames):
            try:
                (tag, dummy, filename, rev) = line.split()
            except ValueError:
                die("tags table in bad format.")
            do_it_or_die("rcs -n%s:%s %s" % (tag, rev, os.path.basename(filename)))

    # Replace all files for which there exist SCCS masters,
    progress("replacing workfiles...")
    for filename in os.listdir("SCCS"):
        if is_sccsmaster(master):
            do_it("co " + rcsfile(workfile(master)))
    progress("done.")

# sccs2rcs ends here

File: tmp/sccs2rcs.pl
#!/pro/bin/perl

use strict;
use warnings;

# Sccs2rcs is a script to convert an existing SCCS history into an RCS
# history without losing any of the information contained therein.
# It has been tested under the following OS's:
#     SunOS 4.1.3 5.6 5.7
#     Linux
#
# Things to note:
#   + It will NOT delete or alter your ./SCCS history under any
#     circumstances.
#
#   + Run in a directory where ./SCCS exists and where you can
#       create ./RCS
#
#   + /usr/local/bin is put in front of the default path.
#     (SCCS under Ultrix is set-uid sccs, bad bad bad, so
#     /usr/local/bin/sccs here fixes that)
#
#   + Date, time, author, comments, branches, are all preserved.
#
#   + If a command fails somewhere in the middle, it bombs with
#     a message -- remove what it's done so far and try again.
#         "rm -rf RCS; sccs unedit `sccs tell`; sccs clean"
#     There is no recovery and exit is far from graceful.
#     If a particular module is hanging you up, consider
#     doing it separately; move it from the current area so that
#     the next run will have a better chance or working.
#     Also (for the brave only) you might consider hacking
#     the s-file for simpler problems:  I've successfully changed
#     the date of a delta to be in sync, then run "sccs admin -z"
#     on the thing.
#
#   + After everything finishes, ./SCCS will be moved to ./old-SCCS.
#
# This file may be copied, processed, hacked, mutilated, and
# even destroyed as long as you don't tell anyone you wrote it.
#
# Ken Cox
# Viewlogic Systems, Inc.
# [EMAIL PROTECTED]
# ...!harvard!cg-atla!viewlog!kenstir
#
# Various hacks made by Brian Berliner before inclusion in CVS contrib area.
# Converted to perl by Michael Sterrett - [EMAIL PROTECTED]
#     Hard to tell if that should be filed under "mutilated" or "destroyed".

############################################################
# Globals
#
my $logfile = "/tmp/sccs2rcs_${$}_log";

# Could have used a hash for the keywords except we want to guarantee
# the order
my @sccs_keywords = (
    '%W%[       ]*%G%',
    '%W%[       ]*%E%',
    '%W%',
    '%Z%%M%[    ]*%I%[  ]*%G%',
    '%Z%%M%[    ]*%I%[  ]*%E%',
    '%M%[       ]*%I%[  ]*%G%',
    '%M%[       ]*%I%[  ]*%E%',
    '%M%',
    '%I%',
    '%G%',
    '%E%',
    '%U%',
    '%P%'
);
# This is used for checking if we need to do keyword substitution or not.
my $big_sccs_pattern = join('|', @sccs_keywords);

# the quotes surround the dollar signs to fool RCS when I check in this script
my @rcs_keywords = (
    '$'.'Id'.'$',
    '$'.'Id'.'$',
    '$'.'Id'.'$',
    '$'.'SunId'.'$',
    '$'.'SunId'.'$',
    '$'.'Id'.'$',
    '$'.'Id'.'$',
    '$'.'RCSfile'.'$',
    '$'.'Revision'.'$',
    '$'.'Date'.'$',
    '$'.'Date'.'$',
    '',
    '$'.'Source'.'$'
);

############################################################
# Subroutines
#
sub error_exit()
{
    print("\n\nDanger!  Danger!\n");
    print("Some command exited with a non-zero exit status.\n");
    print("Log file exists in $logfile.\n\n");
    print("Incomplete history in ./RCS -- remove it\n");
    print("Original unchanged history in ./SCCS\n");
    exit(1);
}

# This routine is used by the Schwartzian Transform in the main loop
sub by_version
{
    my $limit = ($#{$a} > $#{$b}) ? $#{$b} - 1 : $#{$a} - 1;

    for (my $c = 0; $c <= $limit; $c++) {
        my $x = ($a->[$c] <=> $b->[$c]);
        return $x if ($x);
    }
    return ($#{$a} > $#{$b}) ? 1 : -1;
}

# we'll assume the user set up the path correctly
# for the Pmax, /usr/ucb/sccs is suid sccs, what a pain
#   /usr/local/bin/sccs should override /usr/ucb/sccs there
$ENV{'PATH'} = '/usr/local/bin:' . $ENV{'PATH'};

############################################################
# Error checking
#
die("Error: ./ not writable by you.\n") unless ( -w '.' );
die("Error: ./SCCS directory not found.\n") unless ( -d 'SCCS' );
die("Error: ./old-SCCS directory found.\n") if ( -d 'old-SCCS' );
my @tell = `sccs tell`;
die("Error: sccs tell failed.\n") if ($?);
if (@tell) {
    die("Error: ", scalar(@tell), " file(s) out for edit...clean up before 
converting.\n");
}

if ( -d 'RCS' ) {
    print("Warning: RCS directory exists\n");
    opendir(RCS, 'RCS') or die("Couldn't opendir(RCS): $!\n");
    die("Error: RCS directory not empty\n") if (grep { !/^\.\.?$/o} readdir(RCS));
    closedir(RCS);
} else {
    mkdir('RCS', 0777) or die("Error: Failed to mkdir RCS: $!\n");
}

unlink($logfile);

# Sanity check in case the user added/changed things
unless (@sccs_keywords == @rcs_keywords) {
    die("\@sccs_keywords and \@rcs_keywords are mismatched\n");
}

############################################################
# Get some answers from user
#
print("\nDo you want to be prompted for a description of each\n");
print("file as it is checked in to RCS initially?\n");
print("(y=prompt for description, n=null description) [y] ?");
my $prompt_for_desc = (<> =~ /^y?$/io) ? 1 : 0;
print("\nThe default keyword substitutions are as follows and are\n");
print("applied in the order specified:\n");
for (my $c = 0; $c < @sccs_keywords; $c++) {
    print("\t$sccs_keywords[$c]\t==>\t$rcs_keywords[$c]\n");
}
print("\n");
print("Do you want to change them [n] ?");
if (<> =~ /^n?$/io) {
    print("good idea.\n");
} else {
    print("You can't always get what you want.\n");
    print("Edit this script file and change the variables:\n");
    print('    @sccs_keywords', "\n", '    @rcs_keywords', "\n");
    exit 0;
}

############################################################
# Loop over every s-file in SCCS dir
#
# match only "s." files and get rid of the "s." at the beginning of the name
opendir(SCCS, 'SCCS') or die("Couldn't opendir(SCCS): $!\n");
# match only "s." files and get rid of the "s." at the beginning of the name
for my $file (sort (grep { s/^s\.//o } readdir(SCCS))) {
    my $firsttime = 1;
    my %revisions; # hash so we're unique (was sort -u in csh so ...)
    # get all the prs info once to save calling "sccs prs" for *each* revision
    my @sccs_prs = `sccs prs $file` or die("sccs prs $file failed\n");

    # pick out just the revision lines and store the date/time and author
    # for the rev loop below
    for (grep { /^D /o } @sccs_prs) {
        @_ = split(' ', $_, 6);
        push(@{$revisions{$_[1]}}, $_[2] . ' ' . $_[3], $_[4]);
    }
    # remove the old file.  We overwrite to it later so this is non-critical
    unlink($file);

    # Schwartzian transform - see man perlfaq4
    # Sorting SCCS versions...what a pain.
    for my $rev ( map { $_->[$#{$_}] } sort by_version map { [ split(/\./o), $_ ] } 
keys(%revisions)) {
        my ($date, $author) = @{$revisions{$rev}};

        # Y2K fixup for date
        $date =~ s/^(\d\d)/$_ = ($1 + 0 < 70) ? "20$1" : "19$1"/oe or
            die("What the...?!! Bad dates found: $date\n");

        print("\n==> file $file, rev=$rev, date=$date, author=$author\n");

        # add RCS keywords in place of SCCS keywords and create the file
        open(F, ">$file") or die("Couldn't open $file for write: $!\n");
        # This could be faster if we read in the entire file, but that would
        # be bad for small-memory machines.  Better safe than sorry.
        for (`sccs get -k -p -r$rev $file 2>> $logfile`) {
            # only do the *slow* s// loop if we'll actually s// something.
            if (/$big_sccs_pattern/o) {
                study;
                # sed? we don't need no stinking sed - do keyword substitutions
                for (my $c = 0; $c < @sccs_keywords; $c++) {
                    s/$sccs_keywords[$c]/$rcs_keywords[$c]/g;
                    # If there aren't any more left, get out early.
                    !/$big_sccs_pattern/o and last;
                }
            }
            print(F);
        }
        close(F);
        error_exit if ($?); # check the return of the `sccs get...`

        # same output as the csh version for that warm fuzzy feeling
        print("checked out of SCCS\n");
        print("performed keyword substitutions\n");

        # check file into RCS
        if ($firsttime) {
            $firsttime = 0;
            unless ($prompt_for_desc) {
                print("about to do ci\n");
                open(CI, "|ci -f -r$rev -d\"$date\" -w$author -m\"Initial revision\" 
$file >> $logfile 2>&1") or die("ci failed");
                close(CI);
                error_exit if ($?);
                print("initial rev checked into RCS without description\n");
            } else {
                print("\nEnter a brief description of the file $file (end w/ 
Ctrl-D):\n");
                my @description = ();
                while (<>) {
                    push(@description, $_);
                }
                open(CI, "|ci -f -r$rev -d\"$date\" -m\"Initial revision\" -w$author 
$file >> $logfile 2>&1") or die("ci failed");
                print(CI @description) or die("description to ci failed\n");
                close(CI);
                error_exit if ($?);
                print("initial rev checked into RCS\n");
            }
        } else {
            # get RCS lock
            my $lckrev = $rev;
            $lckrev =~ s/\.\d+$//o;

            if ($lckrev =~ /\./o) {
                # need to lock the branch -- it is OK if the lock fails
                system("rcs -l$lckrev $file >> $logfile 2>&1");
            } else {
                # need to lock the trunk -- must succeed
                system("rcs -l $file >> $logfile 2>&1");
                error_exit if ($?);
            }
            print("got lock\n");
            my @old_comments = ();
            my $comments = 0;
            for (@sccs_prs) {
                # skip down to the revision we're looking for..until empty line
                next unless (/^D $rev\s/../^$/o);
                last if (/^$/o); # we found what we came for so get out early
                # once we find ^COMMENTS:, grab lines until we're done.
                $comments = 1, next if (/^COMMENTS:/o && !$comments);
                next unless ($comments);
                push(@old_comments, $_);
            }
            open(CI, "|ci -I -f -r$rev -d\"$date\" -w$author $file >> $logfile 2>&1") 
or die("ci failed\n");
            print(CI @old_comments);
            close(CI);
            print("checked into RCS\n");
        }
    }
}
closedir(SCCS);

############################################################
# Clean up
#
print("cleaning up...\n");
rename('SCCS', 'old-SCCS');
print("===================================================\n");
print("       Conversion Completed Successfully\n\n");
print("         SCCS history now in old-SCCS/\n");
print("===================================================\n");
exit(0);
